### DISCUSSION TITLE
Next JS 14 serving precached layout and page while offline

### DISCUSSION THREAD

**Discussion by jlow2397** [serwist/serwist#141] [Q&A]
Hello! Thank you for providing a great framework for setting up a service worker!

I have gone through the precaching docs [here](https://serwist.pages.dev/docs/serwist/guide/precaching) but I would love some help in trying to leverage the precache into my current Next JS app for offline routing functionality

I noticed that the static js/css files for my dynamic layouts and pages are stored in the `serwist-precache-v2` cache (which is great!), but when I try navigating to a route that is not in the `rscPrefetch` or `rsc` cache while offline, my browser redirects to the chrome-error page

I understand that when routing the react server component is fetched first, and then the static js and css files are fetched in order to load the page, but is there any way to handle this process of serving the static js and css files directly from the precache to perform an offline navigation?

Any help is greatly appreciated, and I would be more than happy to share my learnings with others as this seems to be a common use case of using serwist with Next JS. Thank you!

**Comment by DuCanhGH** [comment]
@jlow2397 

> I understand that when routing the react server component is fetched first, and then the static js and css files are fetched in order to load the page, but is there any way to handle this process of serving the static js and css files directly from the precache to perform an offline navigation?

That your JS/CSS files are directly served from the precache doesn't change at any time. It's actually your pages not being cached that causes this issue.

To fix this, you have to cache your pages before they are visited. This can be done via `window.serwist.messageSW({ type: "CACHE_URLS", urlsToCache: ["/a", "/b"] })`. If you don't use Serwist's service worker API, you may need to process this message yourself, though. 

Caching RSC responses may be a different story, though, as I haven't heard of a standard way to access the pages' RSC content. For example, in Next.js, statically generated pages have their RSC content dumped into `.txt` files, but dynamically generated ones have theirs only accessible via setting some sort of internal headers or search params.

You may also precache your pages, but that isn't recommended if they are dynamically generated as doing that causes them to always be served from the precache until the service worker is updated to not do so.

**Reply by jlow2397** [reply]
Thank you for the quick reply & suggestion!

`window.serwist.messageSW` works great when online (I see that it's adding into the `others` cache storage), but when offline the request fails since it can't handle the URL

error message: Uncaught (in promise) no-response: no-response... at ea._handle... at ea._getResponse

I'd love to dive deeper into trying to access the RSC while offline or some way to leverage the precached javascript files when trying to load a dynamic route, do you have any suggestions for areas to look into regarding this?

**Reply by DuCanhGH** [reply]
@jlow2397 

> but when offline the request fails since it can't handle the URL

Not that it should matter much, should it? Your service worker will fail to precache assets anyway.

Well, if you want to go all out, here is how to "precache" your pages:

```ts
// Just an example. You may also use any utility library you want or
// simply `Promise.all`.
import { parallel } from "radash";

const serwist = new Serwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: defaultCache,
});

const urlsToCache = ["/a", "/b"];

self.addEventListener("install", (event) => {
  event.wailUntil(
    (async () => {
      await parallel(20, urlsToCache, async (url) => {
        // This code, if fails, throws to ensure that the service worker
        // is not installed.
        const request = new Request(url);
        const responsePromise = this.handleRequest({ request, event });
        if (responsePromise) {
          await responsePromise;
        }
      });
    })(),
  );
});

serwist.addEventListeners();
```

> I'd love to dive deeper into trying to access the RSC while offline or some way to leverage the precached javascript files when trying to load a dynamic route, do you have any suggestions for areas to look into regarding this?

Not sure what you mean by "access the RSC while offline". To no one's surprise, a page's RSC content... well, lives in the server, so you can't expect to access it without internet.

**Reply by jlow2397** [reply]
thanks for the snippet but I don't think that I'm looking to "precache" the pages initially, but instead i think the main thing i'm trying to do is deconstruct the Next JS route handler so that it can feed the appropriate javascript layouts & page js files (which are already precached successfully via the service worker) upon an offline navigation (if the user happens to go offline after loading the page)

that may be out of the scope of serwist, but it seems like you may have experience with that since you have worked with the RSC responses before. I would love to write maybe a custom handler to extend this logic if you think it's possible

**Reply by jlow2397** [reply]
One thing I noticed when testing was that an unsuccessful offline navigation resulted in this warning being triggered:

`The FetchEvent for "<URL>" resulted in a network error response: the promise was rejected.` in the handleFetch function

but a successful offline navigation will throw this error after successfully routing: 

`Failed to fetch RSC payload for "<URL>". Falling back to browser navigation.`

For more context, all of these pages are nested within the same client-side layout

**Reply by DuCanhGH** [reply]
@jlow2397 

> thanks for the snippet but I don't think that I'm looking to "precache" the pages initially, but instead i think the main thing i'm trying to do is deconstruct the Next JS route handler so that it can feed the appropriate javascript layouts & page js files (which are already precached successfully via the service worker) upon an offline navigation (if the user happens to go offline after loading the page)

That you need to cache your pages is due to critical pieces of your pages living on the server. In App Router, RSCs are such pieces, and in Pages Router, server data (`getStaticProps`, `getServerSideProps`, etc.). Such pieces are not available to the client bundle, obviously, meaning that your JS files are not enough to render the pages.

> Failed to fetch RSC payload for "<URL>". Falling back to browser navigation.

This is thanks to your pages' HTML payload being cached :)

**Reply by jlow2397** [reply]
Ah thank you for the explanation!

Because of this, I've decided to change our dynamic url structure similar to this: https://github.com/serwist/serwist/issues/49#issuecomment-1904531710

where instead of /thread/\<thread-id\> it's now /thread?id=\<thread-id\>

is there an additional parameter i need to add to the serwist config so that it can successfully cache the /thread page? 

here's what i'm trying right now: 

```
const serwist = new Serwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: defaultCache,
  precacheOptions: {
    ignoreURLParametersMatching: [new RegExp('([^s]+)', 'i')]
  }
})
```

edit: i'm also calling window.serwist.messageSW with the /thread path on a useEffect on intialization which is successfully populating the others cache

**Reply by DuCanhGH** [reply]
@jlow2397 

> where instead of /thread/<thread-id> it's now /thread?id=<thread-id>

Well, if you do that, just make sure that the page doesn't use `searchParams` on the server.

You may precache the page via `additionalPrecacheEntries`, but it may result in your application unexpectedly having multiple versions at the same time. Be sure to set `clientsClaim` and `skipWaiting` to `false`.

It is best that you cache rather than precache them.

**Reply by jlow2397** [reply]
> just make sure that the page doesn't use searchParams on the server.

I'm using the searchParams prop on a client component page, will that affect the performance? If so, what would be a better method to read search parameter data?

I noticed that the network fetch is unique every time the id search parameter changes (which causes the load to fail when offline) so that might be what you are describing?

**Reply by DuCanhGH** [reply]
@jlow2397 

> will that affect the performance?

It is not really about performance, but rather about how the page would have incorrect data if you accessed `searchParams` on the server. Use `useSearchParams` instead.

> I noticed that the network fetch is unique every time the id search parameter changes (which causes the load to fail when offline) so that might be what you are describing?

Use `Strategy#options.matchOptions` like so:

```js
import { NetworkFirst, Serwist } from "serwist";

const serwist = new Serwist({
  runtimeCaching: [
    {
      matcher: ({ request }) => request.url.pathname === "/thread",
      handler: new NetworkFirst({
        matchOptions: {
          ignoreSearch: true,
          ignoreVary: true,
        },
      }),
    },
    ...defaultCache,
  ],
});
```

Or, if you precache the page, use `precacheOptions.ignoreURLParametersMatching`.

**Reply by jlow2397** [reply]
thank you adding a matcher for the request works great to cache the url

when rendering the cached page while offline however, it throws an error and shows the RSC router tree directly on the page like this: 

`0:["7_kj_6nxOoDMXvGYiStPS",[["children",["workspace","hq","d"],"children","mail","children","thread",["thread",{"children":["__PAGE__?{\"id\":\"18f9e4f34317c53a\"}",{}]},"$undefined","$undefined",true],null,null]]]`

is this also related to an incorrect cache? in the network tab it loads the initial page from the service worker with a 200 response

**Reply by jlow2397** [reply]
I recently tried another approach with just a single url /thread and the id passed via a context state variable, but when navigating to the /thread page while offline it's triggering a full page reload (where it re-renders all the shared components in the layout)

is there a setting or caching strategy that is causing that behavior?

**Reply by jlow2397** [reply]
just a quick update: I ended up using next.prefetch('/thread') and it seems to have resolved the issue instead of using serwist.messageSW 

thank you for all your help btw it's been great getting to learn a bit more about service workers through this package!

**Reply by DuCanhGH** [reply]
@jlow2397 

> next.prefetch('/thread')

😅

Alright, if that works in your favour.

> I recently tried another approach with just a single url /thread and the id passed via a context state variable, but when navigating to the /thread page while offline it's triggering a full page reload (where it re-renders all the shared components in the layout)

Not too sure about this. Probably Next.js falling back to an actual browser navigation.

> is this also related to an incorrect cache? in the network tab it loads the initial page from the service worker with a 200 response

Well, maybe. What I suggested was... well, a hypothesis, and I hadn't tried that before suggesting you to do so. In my opinion, caching some specific pages with `themeId` specified is the best option, but you do you :)
